### ISSUE TITLE
std::vector injection?

### ISSUE THREAD

**Issue by nickdademo** [romanpauk/dingo#1]
Thanks for the nice library!

I have the following class which takes a vector of "data":
```
SomeClass::SomeClass(std::vector<Data> data) 
```

I'm trying a basic test with:
```
        dingo::container<> container;
        container.register_type<dingo::scope<dingo::external>, dingo::storage<std::vector<Data>>>(getVectorData());
        container.register_type<dingo::scope<dingo::shared>, dingo::storage<SomeClass>>();
```
which gives errors like the below on MSVC (VS 17.11.5):
```
'dingo::class_instance_resolver': too few template arguments
'dingo::class_instance_resolver': use of class template requires template argument list
```
Am I doing something wrong?

**Comment by romanpauk** [comment]
Hi, the issue you are seeing is most probably caused by missing <dingo/storage/shared.h> include. It is quite puzzling compile error, I'll look at what can be done to make it less error-prone. And worse, some includes got into container.h, so with unique storage there is no issue out of the box, but with shared, there is. As the storage types should be user provided, I've removed the includes from container.h.

Please see a test for external vector here: https://github.com/romanpauk/dingo/blob/2e2f871782af5a169121a26051b1a32d1d3e517b/test/external.cpp#L322

**Comment by nickdademo** [comment]
> Hi, the issue you are seeing is most probably caused by missing <dingo/storage/shared.h> include. It is quite puzzling compile error, I'll look at what can be done to make it less error-prone. And worse, some includes got into container.h, so with unique storage there is no issue out of the box, but with shared, there is. As the storage types should be user provided, I've removed the includes from container.h.
> 
> Please see a test for external vector here:
> 
> https://github.com/romanpauk/dingo/blob/2e2f871782af5a169121a26051b1a32d1d3e517b/test/external.cpp#L322

Thanks Roman.

My includes are as follows:
``` 
#include <dingo/container.h>
#include <dingo/index/array.h>
#include <dingo/storage/shared.h>
```

**Comment by nickdademo** [comment]
Hrmm, what I am doing looks exactly like the test you linked.

**Comment by romanpauk** [comment]
Seems you are missing <dingo/storage/external>.

**Comment by nickdademo** [comment]
> Seems you are missing <dingo/storage/external>.

Yep, exactly! Just tried that now and that's the cause! Thanks!

**Comment by nickdademo** [comment]
Side question: is resolving nested `structs` supported?
i.e. calling `resolve<Outer>()` for the below:
```
struct Outer
{
    struct InnerA
    {
        std::shared_ptr<SomeClass> c;
    } a;

    struct InnerB
    {
        std::unique_ptr<AnotherClass> d;
    } b;
};
```

**Comment by romanpauk** [comment]
Not correctly, it is required to register InnerA/InnerB with the container - https://github.com/romanpauk/dingo/compare/master...feature/nested_resolution. But there already is support for constructing un-managed types through construct<> call, so this is just an generalization of that principle that I think can be implemented.

**Comment by romanpauk** [comment]
Okay, so the code in above branch has some other issues. But the library supports override the factory method used for resolution, so something like this works:

`container.template register_type<scope<unique>, storage<Outer>, factory<constructor<Outer(std::shared_ptr<SomeClass>, std::unique_ptr<AnotherClass>)>>>();`

I've tried some hacky experiments and generally it seems to be possible (somehow works, just breaks some tests here and there). It definitely is an interesting use-case that I've overlooked, so I'll try to handle that, just can't promise when.

**Comment by nickdademo** [comment]
Thanks for investigating Roman!

Definitely would be great to have :) - FYI: `bext-di` supports it out of the box with no extra code.

**Comment by romanpauk** [comment]
Agree, will work on it. I have a four line change that proves that generally there are no issues with supporting unknown types found during resolution, except that:

1) the 4 lines are PoC, there will be some refactoring needed
2) you've helped to uncover previously unknown issue with aggregate initialization of std::unique_ptr
3) with the PoC, gcc 14 ends up compiling infinitely some of the tests/examples

So stay tuned!

**Comment by nickdademo** [comment]
> Agreed, will work on it. I have a four line change that proves that generally there are no issues with supporting unknown types found during resolution, except that:
> 
>     1. the 4 lines are PoC, there will be some refactoring needed
> 
>     2. you've helped to uncover previously unknown issue with aggregate initialization of std::unique_ptr
> 
>     3. with the PoC, gcc 14 ends up compiling infinitely some of the tests/examples
> 
> 
> So stay tuned!

Great to hear! I actually cannot seem to get `dingo` to construct a simple struct of `unique_ptr`s - so maybe that's related to what you have discovered (hopefully!).

**Comment by romanpauk** [comment]
Pushed support for nested resolution, please see the test here: https://github.com/romanpauk/dingo/blob/master/test/nested_resolution.cpp. The unique_ptr/shared_ptr issue goes deeper, I don't have a solution yet.

**Comment by romanpauk** [comment]
Pushed another set of changes, addressing shared_ptr/unique_ptrs. There is a fundamental issue with the way constructor deduction works: before deduction is attempted, one needs to specify if reference (operator T&, operator T&&) or value (operator T) will be used. So for example, is value-based deduction is used, it will correctly deduce non-copyable types (like unique_ptrs) and also copyable types (like shared_ptrs). But if there is a simple reference, this will fail to compile. On the other hand, if reference-based deduction is used, it can initialize copyable types, and it can initialize references, yet it fails with non-copyable types, as it will select deleted constructor. So to detect a constructor argument correctly, we need to know the type of that argument, so we can detect it and that is hard. I think that boost-di, as it is compile time, knows a set of all types to be deduced and is able to work out proper operator T/T&/T&&. Dingo, being runtime-based, can't do it, so the only way out is that users will be able to choose - either the class is to be returned from the container and it does not have any references to other instances (value-based deduction) or the class can use references to existing instances managed by the container (reference-based deduction). If there is a need to mix copyable and non-copyable types in a single class, the constructor can either be wrapped in a macro, or the factory can be fully typed while registering the class with the container. The deduction, without reflection, will always be just a compromise, I am sorry.

**Comment by nickdademo** [comment]
> Pushed another set of changes, addressing shared_ptr/unique_ptrs. There is a fundamental issue with the way constructor deduction works: before deduction is attempted, one needs to specify if reference (operator T&, operator T&&) or value (operator T) will be used. So for example, is value-based deduction is used, it will correctly deduce non-copyable types (like unique_ptrs) and also copyable types (like shared_ptrs). But if there is a simple reference, this will fail to compile. On the other hand, if reference-based deduction is used, it can initialize copyable types, and it can initialize references, yet it fails with non-copyable types, as it will select deleted constructor. So to detect a constructor argument correctly, we need to know the type of that argument, so we can detect it and that is hard. I think that boost-di, as it is compile time, knows a set of all types to be deduced and is able to work out proper operator T/T&/T&&. Dingo, being runtime-based, can't do it, so the only way out is that users will be able to choose - either the class is to be returned from the container and it does not have any references to other instances (value-based deduction) or the class can use references to existing instances managed by the container (reference-based deduction). If there is a need to mix copyable and non-copyable types in a single class, the constructor can either be wrapped in a macro, or the factory can be fully typed while registering the class with the container. The deduction, without reflection, will always be just a compromise, I am sorry.

Thanks for the detailed explanation!
Still cannot manage to construct a simple `struct` of `unique_ptr`s. Do I need to use `constructor_detection`?

**Comment by romanpauk** [comment]
Nope, it still does not work. With some temporary changes I can create a struct of unique_ptrs, but can't create the same with shared_ptrs. The whole unmanaged creation concept or how to call it is not working well, the library evolved from service locator pattern, so its capability to create unmanaged classes is quite lacking. I am afraid current code will be very hard to fix, so perhaps it is time for 4th version, built from the scratch on this unmanaged creation... sorry for the bad news.
